java/arcs/core/data/proto/RecipeProtoDecoder.kt

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.data.proto
+
+import arcs.core.data.ParticleSpec
+import arcs.core.data.Recipe
+import arcs.core.util.Result
+
+/** Converts a [RecipeProto] into [Recipe]. */
+fun RecipeProto.decode(particleSpecs: Map<String, ParticleSpec>): Recipe {
+    val recipeHandles = mutableMapOf<String, Recipe.Handle>()
+    handlesList.forEach {
+        val oldValue = recipeHandles.put(it.name, it.decode())
+        require(oldValue == null) {
```

Comment:
`require` and `requireNotNull` throw `IllegalArgumentException`, while `check` and `checkNotNull` throw `IllegalStateException`.... I'm not sure which would be more appropriate in this context, but I wanted to make you aware of the difference so you can decide.

Reply:
Ah, thanks for the info. `IllegalArgumentException` seems OK here because duplicate handle happens when the receiver `RecipeProto` is not valid.